src/pybind/rbd.pyx

```diff
@@ -0,0 +1,1431 @@
+    """
+    This class represents an RBD image. It is used to perform I/O on
+    the image and interact with snapshots.
+
+    **Note**: Any method of this class may raise :class:`ImageNotFound`
+    if the image has been deleted.
+    """
+    cdef rbd_image_t image
+    cdef bint closed
+    cdef object name
+    cdef object ioctx
+    cdef rados_ioctx_t _ioctx
+
+    def __init__(self, ioctx, name, snapshot=None, read_only=False):
+        """
+        Open the image at the given snapshot.
+        If a snapshot is specified, the image will be read-only, unless
+        :func:`Image.set_snap` is called later.
+
+        If read-only mode is used, metadata for the :class:`Image`
+        object (such as which snapshots exist) may become obsolete. See
+        the C api for more details.
+
+        To clean up from opening the image, :func:`Image.close` should
+        be called.  For ease of use, this is done automatically when
+        an :class:`Image` is used as a context manager (see :pep:`343`).
+
+        :param ioctx: determines which RADOS pool the image is in
+        :type ioctx: :class:`rados.Ioctx`
+        :param name: the name of the image
+        :type name: str
+        :param snapshot: which snapshot to read from
+        :type snaphshot: str
+        :param read_only: whether to open the image in read-only mode
+        :type read_only: bool
+        """
+        name = cstr(name, 'name')
+        snapshot = cstr(snapshot, 'snapshot', opt=True)
+        self.closed = True
+        self.name = name
+        # Keep around a reference to the ioctx, so it won't get deleted
+        self.ioctx = ioctx
+        cdef:
+            rados_ioctx_t _ioctx = convert_ioctx(ioctx)
+            char *_name = name
+            char *_snapshot = opt_str(snapshot)
+        if read_only:
+            with nogil:
+                ret = rbd_open_read_only(_ioctx, _name, &self.image, _snapshot)
+        else:
+            with nogil:
+                ret = rbd_open(_ioctx, _name, &self.image, _snapshot)
+        if ret != 0:
+            raise make_ex(ret, 'error opening image %s at snapshot %s' % (name, snapshot))
+        self.closed = False
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type_, value, traceback):
+        """
+        Closes the image. See :func:`close`
+        """
+        self.close()
+        return False
+
+    def close(self):
+        """
+        Release the resources used by this image object.
+
+        After this is called, this object should not be used.
+        """
+        if not self.closed:
+            self.closed = True
+            with nogil:
+                ret = rbd_close(self.image)
+            if ret < 0:
+                raise make_ex(ret, 'error while closing image %s' % (
+                              self.name,))
+
+    def __del__(self):
+        self.close()
+
+    def __str__(self):
+        s = "rbd.Image(" + dict.__repr__(self.__dict__) + ")"
```

Comment:
That came from the original Python version, but more importantly, it's dumb here, because the Cython version uses `cdef` member variables and I don't think those show up in self.__dict__. I'll fix it up to do something useful tomorrow (I don't have a usable Ceph test environment at home).